mutation/mutation_compactor.hh

```diff
@@ -250,14 +252,36 @@ private:
     }
 
     bool can_purge_tombstone(const tombstone& t, is_shadowable is_shadowable, const gc_clock::time_point deletion_time) {
+        bool purgeable = false;
+        std::optional<max_purgeable::timestamp_source> timestamp_source;
+
         if (_tombstone_gc_state.cheap_to_get_gc_before(_schema)) {
             // if retrieval of grace period is cheap, can_gc() will only be
             // called for tombstones that are older than grace period, in
             // order to avoid unnecessary bloom filter checks when calculating
             // max purgeable timestamp.
-            return satisfy_grace_period(deletion_time) && can_gc(t, is_shadowable);
+            purgeable = satisfy_grace_period(deletion_time);
+            if (purgeable) {
+                purgeable = can_gc(t, is_shadowable, timestamp_source);
+            }
+        } else {
+            purgeable = can_gc(t, is_shadowable, timestamp_source) && satisfy_grace_period(deletion_time);
         }
-        return can_gc(t, is_shadowable) && satisfy_grace_period(deletion_time);
+
+        if constexpr (sstable_compaction()) {
+            if (_tombstone_stats && t) {
+                ++_tombstone_stats->attempts;
+                if (!purgeable && timestamp_source) {
+                    if (timestamp_source.value() == max_purgeable::timestamp_source::memtable_possibly_shadowing_data) {
+                        ++_tombstone_stats->failures_due_to_overlapping_with_memtable;
+                    } else if (timestamp_source.value() == max_purgeable::timestamp_source::other_sstables_possibly_shadowing_data) {
+                        ++_tombstone_stats->failures_due_to_overlapping_with_uncompacting_sstable;
+                    }
+                }
```

Comment:
I have an allergy to stats code using branches. How about making the counters part of an array, and using the timestamp_source as an index into that array?